rfcs/20190208-pybind11.md

```diff
@@ -0,0 +1,327 @@
+3.  binary size
+4.  performance of the Python API
+
+
+## Motivation
+
+
+### What is pybind11?
+
+[Pybind11](https://github.com/pybind/pybind11) is a header-only library that exposes C++ functions and classes to Python to create Python bindings of existing C++ code. Pybind11 is lightweight, and unlike other libraries that serve a similar purpose, it is purely focussed on binding. Pybind11 is relatively new, but is rapidly growing in contributors and adopters. It's an adaptation of [Boost.Python](https://www.boost.org/doc/libs/1_66_0/libs/python/doc/html/index.html) which is quite reputable, but has a lot of extra libraries. Pybind11 is basically the only the binding generation portions of Boost.Python.
+
+
+### Cons with SWIG
+
+1. SWIG auto-generated code is not optimal for performance.
+2.  SWIG needs the swig toolchain.
+3.  SWIG requires .i or .swig files which require knowledge of SWIG directives.
+    1.  We've changed .swig to .i files to get around readability issues
+4.  SWIG combined with C++ is relatively complicated to understand.
+
+
+### Why not CLIF?
+
+1.  The CLIF wrapping directives are more complex [[Link](https://github.com/google/clif/blob/master/clif/python/README.md#api-specification-language)]
+  1.  Need a separate .cliff file
+    2.  Similar to SWIG, adds a toolchain dependency beyond just the C++ compiler (and adds another dependency on a different subset of LLVM)
+2.  Most literature is optimized for cmake as opposed to bazel
+3.  Fewer developers and adopters in open source
+    1.  [No activity](https://github.com/google/clif) for nearly a year
+
+
+### Pro/Cons with pybind11
+
+
+| Pros: | Cons: |
+|-------|-------|
+| No need to link against any additional libraries|Less support|
+|Function signatures are precomputed at compile time which leads to smaller binaries|Our cases may be more complicated than their examples|
+|No auto-generated code|Heavy reliance on another third party library|
+|Much simpler and easier to debug||
+
+
+## Design Proposal
+
+We should deprecate SWIG and replace the wrapper with pybind11. It would be impractical to create a single cl where we transfer everything so we will do a transition in place. Each [.i file](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tensorflow.i) is a swig file that has directives to wrap and export various C++ classes and functions. There are 60 files across all of TensorFlow. On average there are roughly 20 symbols per file. We can do a rolling transition where we export certain files twice. We can export some symbols with SWIG and the remaining ones with pybind. 
+
+
+### Impact
+
+We can get rid of roughly 60 bloated SWIG files by transitioning to pybind11. We can also make documentation and exporting functions much easier. We can save countless hours on adding classes and ops by making exporting them to Python extremely easy. 
+
+1.  Faster build times by decreasing code generation and simply exposing the code directly.
```

Comment:
Interestingly, the build times reported in the LSST blog post (see the "Issues" section below) are comparable between pybind11/SWIG versions. Do you have another reference where the build times are consistently faster?

Reply:
I don't have an official metrics. I will definitely document the differences as I roll out implementation. We are making the assumption based on getting rid of a few dependency links in the middle (Swig generated python file, pywrap_tensorflow.py) between C++ and the python function call.